text/0003-templates.md

```diff
@@ -0,0 +1,258 @@
+
+A few of the problems above can be mitigated by rendering the template package itself before publishing, but this introduces other problems. Toolkit templates install dependencies in the referencing package as `devDependencies`, which isn't desirable in other templates.
+
+The existing design is inspired from the concept of "traits" in some programming languages, which provide reusability without inheritance.
+
+### Specifiers
+
+Templates can be specified through:
+
+1. The `name@version` syntax to refer to an external template fetched from a registry.
+2. The `name@workspace:` syntax to refer to a template that's also a workspace package.
+3. The `file:<path>` syntax to refer to an on-disk path.
+
+## Rationale and Alternatives
+
+### Syncpack
+
+[Syncpack](https://github.com/JamieMason/syncpack/) is a great open source tool for keeping `package.json` dependency specifiers in sync on disk.
+
+The proposed solution allows metadata to be defined in a singular file without copying definitions to other files on disk. This is a capability only possible by the package manager reading `package.json` files.
+
+### Comparison to overrides/resolutions
+
+An alternative mechanism for the version catalog is the [`pnpm.overrides` feature](https://pnpm.io/package_json#pnpmoverrides). While mechanically this allows you to set the version of a dependency across all workspace packages, it can be a bit unexpected when if `pnpm.overrides` rewrites a dependency's dependency to an incompatible version silently.
+
+`pnpm.overrides` is ultimately intended for a different purpose. The NPM RFC for a similar feature explicitly states that it should be used as a short-term hack to fix vendor problems.
+
+> Using this feature should be considered a hack in most cases, something that is done temporarily while waiting for a bug to be fixed, or to avoid excessive duplication caused by an overly strict meta-dependency specifier.
+https://github.com/npm/rfcs/blob/main/accepted/0036-overrides.md
+
+The `catalog:` protocol is conversely intended for long-lived usage.
+
+## Implementation
+
+### Fetching
+
+Templates are not installed as standard dependencies and linked into `node_modules`. If a package refers to an external template on an NPM registry, only the metadata will be fetched. The fetch will be performed [without the abbreviated header](https://github.com/npm/registry/blob/master/docs/responses/package-metadata.md#abbreviated-metadata-format), so the full document is retrieved.
+
+If only a subset of the version catalog is used, the full catalog of dependencies does not need to be installed.
+
+### Lockfile
+
+No explicit changes will be made to the `pnpm-lock.yaml` file. Any `importers` entries referencing a template will have their rendered result saved to the lockfile.
```

Comment:
If we don't save the template versions to the lockfile, we'll have to fetch all template metadata files in order to verify  that the lockfile is up-to-date. Which might be OK. I don't know.

But also, currently we store the integrity checksum of every package. Should store the integrity checksums of the template packages? What if someone hijacks a template package?

Reply:
> If we don't save the template versions to the lockfile, we'll have to fetch all template metadata files in order to verify that the lockfile is up-to-date. Which might be OK. I don't know.

I'm also a bit unsure. My thought was that the template will be saved to the pnpm content addressable store. (I'm realizing I should have elaborated on that in the RFC.) From there, loading the template subsequent times becomes a disk block I/O operation, much like reading part of `pnpm-lock.yaml` would.

> But also, currently we store the integrity checksum of every package. Should store the integrity checksums of the template packages? What if someone hijacks a template package?

This is a great point. We should absolutely store the integrity checksums of the template packages. Thank you for the catch. Will update the RFC.

Reply:
Commit cb35007544bce664af0387ff93d93bcb7a194275 now describes a new `templates` block.